### PR TITLE
[ViMP] Support playlists, add more instances

### PR DESCRIPTION
**Pull request by FestplattenSchnitzel**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

- [x] Support playlists (albums, categories, channels)
- [x] Extract thumbnails
- [x] Add more sites

### CODE REVIEW

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 1]
```diff
@@ -162,7 +181,86 @@ def _real_extract(self, url):
...
+            # Request returns last n video URLs if existing pages have been "used up"
+            if urls[-1:] != current_urls[-1:]:
+                urls += current_urls
+                i += 1
+            else:
+                break
+
+        return self.playlist_result(
+            (self.url_result(host + url, VideocampusSachsenIE) for url in urls),
+            playlist_title=title, id=f'{mode}-{content_id}')
```
Comment (FestplattenSchnitzel) [inline yt_dlp/extractor/videocampus_sachsen.py:266]: ViMP has three "playlist modes": albums, categories, channels. Each of them has it's own set of identifiers (both `album-1` and `channel-1` can exist).

Is it OK to self-buid the `id` here? Should it be constructed differently?

Comment (pukkandan) [reply yt_dlp/extractor/videocampus_sachsen.py:266]: playlist ids don't matter much, as long as they are sensible enough

**Comment by FestplattenSchnitzel** [comment]
Actually `site-request` does not apply here, but `site-enhancement`. I only added more `ViMP` instances.

I apologise if the wording caused confusion.

**Comment by pukkandan** [comment]
It's just that I label new extractors as "site request" even if other parts of the same website are already supported

File: yt_dlp/extractor/_extractors.py [hunk 2]
```diff
@@ -1917,7 +1917,10 @@
 from .vidbit import VidbitIE
 from .viddler import ViddlerIE
 from .videa import VideaIE
-from .videocampus_sachsen import VideocampusSachsenIE
+from .videocampus_sachsen import (
+        VideocampusSachsenIE,
+        ViMPPlaylistIE,
+)
```
Comment (pukkandan) [inline yt_dlp/extractor/_extractors.py:1923]: wrong indentation

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 3]
```diff
@@ -2,18 +2,25 @@
 
 from .common import InfoExtractor
 from ..compat import compat_HTTPError
-from ..utils import ExtractorError
+from ..utils import (
+    ExtractorError,
+    urlencode_postdata,
+)
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:8]:
```suggestion
from ..utils import ExtractorError, urlencode_postdata
```

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 4]
```diff
@@ -162,7 +181,86 @@ def _real_extract(self, url):
...
-            'subtitles': subtitles
+            'subtitles': subtitles,
         }
+
+
+class ViMPPlaylistIE(InfoExtractor):
+    IE_NAME = 'ViMP:Playlist'
+    _VALID_URL = r'''(?x)(?P<host>https?://(?:%s))/(?:
+        album/view/aid/(?P<album_id>[0-9]+)|(?P<c_mode>category|channel)/(?P<c_name>[\w-]+)/(?P<c_id>[0-9]+)
+        )''' % ('|'.join(map(re.escape, VideocampusSachsenIE._INSTANCES)))
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:195]: a bit hard to read
```suggestion
    _VALID_URL = r'''(?x)(?P<host>https?://(?:%s))/(?:
        album/view/aid/(?P<album_id>[0-9]+)|
        (?P<mode>category|channel)/(?P<name>[\w-]+)/(?P<id>[0-9]+)
    )''' % '|'.join(map(re.escape, VideocampusSachsenIE._INSTANCES))
```
I also changed the group names. Change below accordingly

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 5]
```diff
@@ -162,7 +181,86 @@ def _real_extract(self, url):
...
+            'vars[context_id]': content_id,
+            'vars[layout]': 'thumb',
+            'vars[per_page][thumb]': '10',
+        }
+
+        urls, i = [], 1
+        while True:
+            current_page = self._download_webpage(
+                f'{host}/media/ajax/component/boxList/{url_part}?page={i}&page_only=1',
+                content_id, data=urlencode_postdata(data))
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:254]:
```suggestion
                f'{host}/media/ajax/component/boxList/{url_part}', content_id,
                query={'page': i, 'page_only': 1}, data=urlencode_postdata(data))
```

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 6]
```diff
@@ -162,7 +181,86 @@ def _real_extract(self, url):
...
+            current_urls = re.findall(r'"(.+?/video/.+?)"', current_page)
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:255]: Avoid non-greedy match. `.+?` => `[^"]+`

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 7]
```diff
@@ -162,7 +181,86 @@ def _real_extract(self, url):
...
+
+            # Request returns last n video URLs if existing pages have been "used up"
+            if urls[-1:] != current_urls[-1:]:
+                urls += current_urls
+                i += 1
+            else:
+                break
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:262]: this should be a PagedList

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 8]
```diff
@@ -162,7 +181,86 @@ def _real_extract(self, url):
...
+            url_part = f'{url_parts[0]}/{c_name}/{url_parts[1]}/{c_id}'
+
+        data = {
+            'vars[mode]': mode,
+            f'vars[{mode}]': content_id,
+            'vars[context]': mode_id,
+            'vars[context_id]': content_id,
+            'vars[layout]': 'thumb',
+            'vars[per_page][thumb]': '10',
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:248]:
```suggestion
        url_part = (f'aid/{album_id}' if album_id
                    else f'category/{c_name}/category_id/{c_id}' if c_mode == 'category'
                    else f'title/{c_name}/channel/{c_id}')
        c_mode = c_mode or 'album'
        data = {
            'vars[mode]': c_mode,
            f'vars[{c_mode}]': album_id or c_id,
            'vars[context]': '3' if album_id else '1' if c_mode == 'category' else '4',
            'vars[context_id]': album_id or c_id,
            'vars[layout]': 'thumb',
            'vars[per_page][thumb]': '10',
        }
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
Since there are no requested changes in the video extractor, I have cherrypicked that portion

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 9]
```diff
@@ -183,3 +184,72 @@ def _real_extract(self, url):
...
             'subtitles': subtitles,
         }
+
+
+class ViMPPlaylistIE(InfoExtractor):
+    IE_NAME = 'ViMP:Playlist'
+    _VALID_URL = r'''(?x)(?P<host>https?://(?:%s))/(?:
+        album/view/aid/(?P<album_id>[0-9]+)|
+        (?P<mode>category|channel)/(?P<name>[\w-]+)/(?P<id>[0-9]+)
+        )''' % ('|'.join(map(re.escape, VideocampusSachsenIE._INSTANCES)))
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:194]:
```suggestion
    )''' % '|'.join(map(re.escape, VideocampusSachsenIE._INSTANCES))
```

File: yt_dlp/extractor/videocampus_sachsen.py [hunk 10]
```diff
@@ -183,3 +184,72 @@ def _real_extract(self, url):
...
+        webpage = self._download_webpage(url, album_id or id, fatal=False) or ''
+        title = (self._html_search_meta('title', webpage, fatal=False)
+                 or self._html_extract_title(webpage))
+
+        url_part = (f'aid/{album_id}' if album_id
+                    else f'category/{name}/category_id/{id}' if mode == 'category'
+                    else f'title/{name}/channel/{id}')
+
+        mode = mode or 'album'
+
```
Comment (pukkandan) [inline yt_dlp/extractor/videocampus_sachsen.py:242]:
```suggestion
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Pull request merged**